tchannel/tchannel.py

```diff
@@ -176,6 +177,16 @@ def call(self, scheme, service, arg1, arg2=None, arg3=None,
         raise gen.Return(result)
 
     def listen(self, port=None):
+        if self._dep_tchannel.is_listening():
+            listening_port = int(self._dep_tchannel.hostport.rsplit(":")[1])
+            if port and port != listening_port:
```

Comment:
So this doesn't make it fully thread-safe. The implementation of `dep_tchannel.listen` is:

    if self.is_listening: raise
    # start listening
    self._server = server  # < this is what sets is_listening to True

So there's this huge area in the middle where you could end up listening twice. You want to acquire a lock or something the first thing when listen() is called so that if another listen has to wait for that to finish before it can check is_listening.